### ISSUE TITLE
race condition in current_line_blame

### ISSUE THREAD

**Issue by jdrouhard** [lewis6991/gitsigns.nvim#312]
If git blame takes a bit of time (common in large codebases with lots of history), it's possible to delete a buffer between the call to git blame and the point that the coroutinue actually attempts to set the extmark.

I think there needs to be another check after the coroutine calls back to the `update()` function after the git blame is complete to ensure the buffer still exists.

https://github.com/lewis6991/gitsigns.nvim/blob/main/lua/gitsigns/current_line_blame.lua#L45-L50:

```lua
   local result = bcache.git_obj:run_blame(buftext, lnum)

   scheduler() -- buffer is deleted at some point between running blame and getting back here

   M.reset(bufnr)
   api.nvim_buf_set_extmark(bufnr, namespace, lnum - 1, 0, {
```

I hit this today and  got this error message:

> Error executing vim.schedule lua callback: ...ck/packer/start/plenary.nvim/lua/plenary/async/async.lua:14: The coroutine failed with this message: .../start/gitsigns.nvim/lua/gitsigns/current_line_blame.lua:50: line value outside range

**Comment by lewis6991** [comment]
I don't think it is because of the buffer being deleted because otherwise you will get a `Invalid buffer id` error instead.

I think the only reason the error can appear is if the buffer is still open, but the line number the extmark is being applied to no longer exists.

I've opened #317 to check that the cursor hasn't moved before applying the extmark. This should at the very least avoid the error you are seeing.

Could you try it out?

**Comment by jdrouhard** [comment]
Thanks, I'll check it out soon. I should mention that I got that error when deleting buffers, not when I was deleting lines, so that error might be generic and apply even if the buffer itself is not longer loaded.

It looks like the fix will mostly work, but might have a gap if the line number of the cursor in the new buffer happens to be exactly the same as the line number of the cursor on the buffer the git blame ran on. Should the new function check the line number and the buffer number or something?

**Comment by lewis6991** [comment]
> Should the new function check the line number and the buffer number or something?

Good point, it probably should.

**Comment by lewis6991** [comment]
Do you have a way to reliably produce the issue or does it just happen randomly? I tried adding arbitrary delays but couldn't get any error.

**Comment by jdrouhard** [comment]
It was seemingly random.

When I get a chance I'll see if putting a git wrapper in my `PATH` that just sleeps for a bit on blame does the trick.
